src/device.hpp

```diff
@@ -217,13 +217,32 @@ struct cvk_device : public _cl_device_id,
         return size * percentage_of_available_memory_reported;
     }
 
-    uint64_t memory_size() const {
-        return std::min(max_alloc_size() * 4, actual_memory_size());
-    }
+    uint64_t max_mem_alloc_size() const {
+        // Min memory as per the specs.
+        auto specMinAllocSz =
+            std::max(std::min((uint64_t)(1024 * 1024 * 1024),
+                              (uint64_t)((1 / 4) * global_mem_size())),
+                     (uint64_t)(32 * 1024 * 1024));
+
+        // Max memory allocation for single buffer is set in config.def to 1024
+        // mb - minimum required by spec. For single allocation this value can
+        // be adjusted with environment variable CLVK_MEM_MAX_ALLOC_SIZE_MB For
+        // multiple allocations(total memory allocations), environment var
+        // CLVK_PERCENTAGE_OF_AVAILABLE_MEMORY_REPORTED can be adjusted.
+        auto maxAllocSz =
+            std::min(m_maintenance3_properties.maxMemoryAllocationSize,
+                     (uint64_t)config.max_mem_alloc_size_mb() * 1024 * 1024);
```

Comment:
I think we should handle the case where the maintenance3 extension is not present too. Value-initialising `m_maintenance3_properties` would be good enough but maybe a bit subtle to reason about. I suggest you either do that or check for the extension's presence.

Reply:
IIRC we tried on a device (hacked our device that we dont support this extension), and made sure we gracefully handled this case. I remember in our local testing, it was gracefully handled. 
But, will look at your suggestion and change accordingly.

Reply:
This test does not mean much as its outcome depends on the value of an uninitialised memory location. If you run the same test under valgrind it'll complain about this.